spec/unit/ssl/ssl_provider_spec.rb

```diff
@@ -0,0 +1,341 @@
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "The issuer '/CN=Test CA Subauthority' of certificate '/CN=signed' cannot be found locally")
+    end
+
+    it 'raises if root issuer is missing' do
+      expect {
+        subject.create_context(config.merge(cacerts: [cert('intermediate.pem')]))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "The issuer '/CN=Test CA' of certificate '/CN=Test CA Subauthority' is missing")
+    end
+
+    it 'raises if cert is not valid yet', unless: Puppet::Util::Platform.jruby? do
+      client_cert.not_before = Time.now + (5 * 60 * 60)
+      expect {
+        subject.create_context(config.merge(client_cert: client_cert))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "The certificate '/CN=signed' is not yet valid, verify time is synchronized")
+    end
+
+    it 'raises if cert is expired', unless: Puppet::Util::Platform.jruby? do
+      client_cert.not_after = Time.at(0)
+      expect {
+        subject.create_context(config.merge(client_cert: client_cert))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "The certificate '/CN=signed' has expired, verify time is synchronized")
+    end
+
+    it 'raises if crl is not valid yet', unless: Puppet::Util::Platform.jruby? do
+      future_crls = global_crls
+      # invalidate the CRL issued by the root
+      future_crls.first.last_update = Time.now + (5 * 60 * 60)
+
+      expect {
+        subject.create_context(config.merge(crls: future_crls))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "The CRL issued by '/CN=Test CA' is not yet valid, verify time is synchronized")
+    end
+
+    it 'raises if crl is expired', unless: Puppet::Util::Platform.jruby? do
+      past_crls = global_crls
+      # invalidate the CRL issued by the root
+      past_crls.first.next_update = Time.at(0)
+
+      expect {
+        subject.create_context(config.merge(crls: past_crls))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "The CRL issued by '/CN=Test CA' has expired, verify time is synchronized")
+    end
+
+    it 'raises if the root CRL is missing' do
+      expect {
+        subject.create_context(config.merge(crls: [crl('intermediate-crl.pem')], revocation: :chain))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "The CRL issued by '/CN=Test CA' is missing")
+    end
+
+    it 'raises if the intermediate CRL is missing' do
+      expect {
+        subject.create_context(config.merge(crls: [crl('crl.pem')]))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "The CRL issued by '/CN=Test CA Subauthority' is missing")
+    end
+
+    it "doesn't raise if the root CRL is missing and we're just checking the leaf" do
+      crls = [crl('intermediate-crl.pem')]
+      subject.create_context(config.merge(crls: crls, revocation: :leaf))
```

Comment:
Nit: any reason for the different structure vs the :chain test above?

Reply:
If we're checking revocation for the leaf-most CA, then puppet should not require the root CA's CRL. If revocation is `:chain`, then we do require all CRLs to be present (one for each CA cert in the chain). So here we don't expect an exception, while above we do.